File: packages/typescript/src/Process.ts
import {
  Command,
  ExecuteResponse,
  Session,
  SessionExecuteRequest,
  SessionExecuteResponse,
  ToolboxApi,
  Workspace,
} from '@daytonaio/api-client'
import { WorkspaceCodeToolbox } from './Workspace'
import { parseApiError } from './utils/errors';

/**
 * Handles process and code execution within a workspace
 * @class Process
 */
export class Process {
  constructor(
    private readonly codeToolbox: WorkspaceCodeToolbox,
    private readonly toolboxApi: ToolboxApi,
    private readonly instance: Workspace,
  ) {}

  /**
   * Executes a shell command in the workspace
   * @param {string} command - Command to execute
   * @returns {Promise<ExecuteResponse>} Command execution results
   */
  public async executeCommand(
    command: string,
    cwd?: string,
    timeout?: number
  ): Promise<ExecuteResponse> {
    let response;
    try {
      response = await this.toolboxApi.executeCommand(this.instance.id, {
        command,
        timeout,
        cwd,
    })
    } catch (error) {
      throw new Error(`Failed to execute command: ${parseApiError(error)}`)
    }
    return response.data
  }

  /**
   * Executes code in the workspace using the appropriate language runtime
   * @param {string} code - Code to execute
   * @returns {Promise<ExecuteResponse>} Code execution results
   */
  public async codeRun(code: string): Promise<ExecuteResponse> {
    const runCommand = this.codeToolbox.getRunCommand(code)

    let response;
    try {
      response = await this.toolboxApi.executeCommand(this.instance.id, {
        command: runCommand,
      })
    } catch (error) {
      throw new Error(`Failed to execute code: ${parseApiError(error)}`)
    }

    return response.data
  }

  /**
   * Creates a new exec session in the workspace
   * @param {string} sessionId - Unique identifier for the session
   * @returns {Promise<ExecuteResponse>} Code execution results
   */
  public async createSession(sessionId: string): Promise<void> {
    try {
      await this.toolboxApi.createSession(this.instance.id, {
        sessionId,
      })
    } catch (error) {
      throw new Error(`Failed to create session: ${parseApiError(error)}`)
    }
  }

  /**
   * Executes a command in the session
   * @param {string} sessionId - Unique identifier for the session
   * @param {SessionExecuteRequest} req - Command to execute and async flag
   * @returns {Promise<SessionExecuteResponse>} Command execution results
   */
  public async executeSessionCommand(sessionId: string, req: SessionExecuteRequest): Promise<SessionExecuteResponse> {
    let response;
    try {
      response = await this.toolboxApi.executeSessionCommand(this.instance.id, sessionId, req)
    } catch (error) {
      throw new Error(`Failed to execute session command: ${parseApiError(error)}`)
    }

    return response.data
  }

  /**
   * Gets the logs for a command in the session
   * @param {string} sessionId - Unique identifier for the session
   * @param {string} commandId - Unique identifier for the command
   * @returns {Promise<string>} Command logs
   */
  public async getSessionCommandLogs(sessionId: string, commandId: string): Promise<string> {
    let response;
    try {
      response = await this.toolboxApi.getSessionCommandLogs(this.instance.id, sessionId, commandId)
    } catch (error) {
      throw new Error(`Failed to get session command logs: ${parseApiError(error)}`)
    }
    return response.data
  }

  /**
   * Gets the session
   * @param {string} sessionId - Unique identifier for the session
   * @returns {Promise<Session>} Session
   */
  public async getSession(sessionId: string): Promise<Session> {
    let response;
    try {
      response = await this.toolboxApi.getSession(this.instance.id, sessionId)
    } catch (error) {
      throw new Error(`Failed to get session: ${parseApiError(error)}`)
    }
    return response.data
  }

  /**
   * Gets the session command
   * @param {string} sessionId - Unique identifier for the session
   * @param {string} commandId - Unique identifier for the command
   * @returns {Promise<Command>} Session command
   */
  public async getSessionCommand(sessionId: string, commandId: string): Promise<Command> {
    let response;
    try {
      response = await this.toolboxApi.getSessionCommand(this.instance.id, sessionId, commandId)
    } catch (error) {
      throw new Error(`Failed to get session command: ${parseApiError(error)}`)
    }
    return response.data
  }

  /**
   * Lists all sessions in the workspace
   * @returns {Promise<Session[]>} List of sessions
   */
  public async listSessions(): Promise<Session[]> {
    let response;
    try {
      response = await this.toolboxApi.listSessions(this.instance.id)
    } catch (error) {
      throw new Error(`Failed to list sessions: ${parseApiError(error)}`)
    }
    return response.data
  }

  /**
   * Deletes a session in the workspace
   * @param {string} sessionId - Unique identifier for the session
   */
  public async deleteSession(sessionId: string): Promise<void> {
    try {
      await this.toolboxApi.deleteSession(this.instance.id, sessionId)
    } catch (error) {
      throw new Error(`Failed to delete session: ${parseApiError(error)}`)
    }
  }
}

File: packages/python/src/daytona_sdk/process.py
"""
Process and code execution within a Daytona workspace.

This module provides functionality for executing commands and running code
in the workspace environment.
"""

from typing import Optional, List
from daytona_api_client import (
    Workspace as WorkspaceInstance,
    ToolboxApi,
    ExecuteResponse,
    ExecuteRequest,
    Session,
    SessionExecuteRequest,
    SessionExecuteResponse,
    CreateSessionRequest,
    Command
)

from daytona_sdk._utils.errors import parse_api_error
from .code_toolbox.workspace_python_code_toolbox import WorkspacePythonCodeToolbox


class Process:
    """Handles process and code execution within a workspace.
    
    Args:
        code_toolbox: Language-specific code execution toolbox
        toolbox_api: API client for workspace operations
        instance: The workspace instance
    """

    def __init__(
        self,
        code_toolbox: WorkspacePythonCodeToolbox,
        toolbox_api: ToolboxApi,
        instance: WorkspaceInstance,
    ):
        self.code_toolbox = code_toolbox
        self.toolbox_api = toolbox_api
        self.instance = instance

    def exec(self, command: str, cwd: Optional[str] = None, timeout: Optional[int] = None) -> ExecuteResponse:
        """Executes a shell command in the workspace.
        
        Args:
            command: Command to execute
            cwd: Working directory for command execution (optional)
            timeout: Optional timeout in seconds
            
        Returns:
            Command execution results
        """
        execute_request = ExecuteRequest(
            command=command,
            cwd=cwd,
            timeout=timeout
        )
        
        try:
            return self.toolbox_api.execute_command(
                workspace_id=self.instance.id,
                execute_request=execute_request
            )
        except Exception as e:
            raise Exception(f"Failed to execute command: {parse_api_error(e)}") from None

    def code_run(self, code: str) -> ExecuteResponse:
        """Executes code in the workspace using the appropriate language runtime.
        
        Args:
            code: Code to execute
            
        Returns:
            Code execution results
        """
        command = self.code_toolbox.get_run_command(code)
        return self.exec(command)

    def create_session(self, session_id: str) -> None:
        """Creates a new exec session in the workspace.
        
        Args:
            session_id: Unique identifier for the session
        """
        request = CreateSessionRequest(sessionId=session_id)
        try:
            self.toolbox_api.create_session(
                workspace_id=self.instance.id,
                create_session_request=request
            )
        except Exception as e:
            raise Exception(f"Failed to create session: {parse_api_error(e)}") from None

    def get_session(self, session_id: str) -> Session:
        """Gets a session in the workspace.
        
        Args:
            session_id: Unique identifier for the session
            
        Returns:
            Session
        """
        try:
            return self.toolbox_api.get_session(
                workspace_id=self.instance.id,
                session_id=session_id
            )
        except Exception as e:
            raise Exception(f"Failed to get session: {parse_api_error(e)}") from None
    
    def get_session_command(self, session_id: str, command_id: str) -> Command:
        """Gets a command in the session.
        
        Args:
            session_id: Unique identifier for the session
            command_id: Unique identifier for the command
            
        Returns:
            Command
        """
        try:
            return self.toolbox_api.get_session_command(
                workspace_id=self.instance.id,
                session_id=session_id,
                command_id=command_id
            )
        except Exception as e:
            raise Exception(f"Failed to get session command: {parse_api_error(e)}") from None

    def execute_session_command(self, session_id: str, req: SessionExecuteRequest) -> SessionExecuteResponse:
        """Executes a command in the session.
        
        Args:
            session_id: Unique identifier for the session
            req: Command to execute and async flag
            
        Returns:
            Command execution results
        """
        try:
            return self.toolbox_api.execute_session_command(
                workspace_id=self.instance.id,
                session_id=session_id,
                session_execute_request=req
            )
        except Exception as e:
            raise Exception(f"Failed to execute session command: {parse_api_error(e)}") from None

    def get_session_command_logs(self, session_id: str, command_id: str) -> str:
        """Gets the logs for a command in the session.
        
        Args:
            session_id: Unique identifier for the session
            command_id: Unique identifier for the command
            
        Returns:
            Command logs
        """
        try:
            return self.toolbox_api.get_session_command_logs(
                workspace_id=self.instance.id,
                session_id=session_id,
                command_id=command_id
            )
        except Exception as e:
            raise Exception(f"Failed to get session command logs: {parse_api_error(e)}") from None

    def list_sessions(self) -> List[Session]:
        """Lists all sessions in the workspace.
        
        Returns:
            List of sessions
        """
        try:
            return self.toolbox_api.list_sessions(
                workspace_id=self.instance.id
            )
        except Exception as e:
            raise Exception(f"Failed to list sessions: {parse_api_error(e)}") from None

    def delete_session(self, session_id: str) -> None:
        """Deletes a session in the workspace.
        
        Args:
            session_id: Unique identifier for the session
        """
        try:
            self.toolbox_api.delete_session(
                workspace_id=self.instance.id,
                session_id=session_id
            )
        except Exception as e:
            raise Exception(f"Failed to delete session: {parse_api_error(e)}") from None

    